contrib/log_correlation/log4j/src/main/java/io/opencensus/contrib/logcorrelation/log4j/OpenCensusTraceContextDataInjector.java

```diff
@@ -0,0 +1,219 @@
+      "io.opencensus.contrib.logcorrelation.log4j."
+          + "OpenCensusTraceContextDataInjector.spanSelection";
+
+  private final Tracer tracer = Tracing.getTracer();
+
+  private final SpanSelection spanSelection;
+
+  /**
+   * How to decide whether to add tracing data from the current span to a log entry.
+   *
+   * @since 0.16
+   */
+  public enum SpanSelection {
+
+    /**
+     * Never add tracing data to log entries. This constant disables the log correlation feature.
+     *
+     * @since 0.16
+     */
+    NO_SPANS,
+
+    /**
+     * Add tracing data to a log entry iff the current span is sampled.
+     *
+     * @since 0.16
+     */
+    SAMPLED_SPANS,
+
+    /**
+     * Always add tracing data to log entries, even when the current span is not sampled. This is
+     * the default.
+     *
+     * @since 0.16
+     */
+    ALL_SPANS
+  }
+
+  /**
+   * Returns the {@code SpanSelection} setting for this instance.
+   *
+   * @return the {@code SpanSelection} setting for this instance.
+   * @since 0.16
+   */
+  public SpanSelection getSpanSelection() {
+    return spanSelection;
+  }
+
+  /**
+   * Constructor to be called by Log4j.
+   *
+   * <p>This constructor looks up the {@link SpanSelection} using the system property {@link
+   * #SPAN_SELECTION_PROPERTY_NAME}.
+   *
+   * @since 0.16
+   */
+  public OpenCensusTraceContextDataInjector() {
+    this(lookUpSpanSelectionProperty());
+  }
+
+  private OpenCensusTraceContextDataInjector(SpanSelection spanSelection) {
+    this.spanSelection = spanSelection;
+  }
+
+  private static SpanSelection lookUpSpanSelectionProperty() {
+    String spanSelectionProperty = System.getProperty(SPAN_SELECTION_PROPERTY_NAME);
+    return spanSelectionProperty == null || spanSelectionProperty.isEmpty()
+        ? DEFAULT_SPAN_SELECTION
+        : parseSpanSelection(spanSelectionProperty);
+  }
+
+  private static SpanSelection parseSpanSelection(String spanSelection) {
+    try {
+      return SpanSelection.valueOf(spanSelection);
+    } catch (IllegalArgumentException e) {
+      return DEFAULT_SPAN_SELECTION;
+    }
+  }
+
+  // The implementation of this method is based on the example in the Javadocs for
+  // ContextDataInjector.injectContextData.
+  @Override
+  public StringMap injectContextData(@Nullable List<Property> properties, StringMap reusable) {
+    if (properties == null || properties.isEmpty()) {
+      return rawContextData();
+    }
+    // Context data has precedence over configuration properties.
+    ThreadContextDataInjector.copyProperties(properties, reusable);
+    reusable.putAll(rawContextData());
+    return reusable;
+  }
+
+  // This method avoids getting the current span when the feature is disabled, for efficiency.
+  @Override
+  public StringMap rawContextData() {
+    switch (spanSelection) {
+      case NO_SPANS:
+        return getContextData();
+      case SAMPLED_SPANS:
+        SpanContext spanContext = tracer.getCurrentSpan().getContext();
```

Comment:
iotw imagine if this were reading directly from the grpc context key

Reply:
I'm not sure I understand.  Do you mean that it would be better to get the tracing data directly from the gRPC context, i.e., `ContextUtils.CONTEXT_SPAN_KEY.get().getContext()`?  What is the benefit of using the context directly?  I thought that a statically accessed tracer would have the same issues as a statically accessed context.

Reply:
I think we are caching a reference to the tracer, which means there's bootstrap ordering concerns. looking up just the key at runtime at least solves that right?

Reply:
sorry I suppose I was unclear.. I mean to say feels like we have two choices:

* we formalize this is supported for any impl that uses CONTEXT_SPAN_KEY (including the default impl), then use this at runtime.

* we consider initialization racing, make a comment why it isn't a concern to cache Tracing.getTracer() when it is likely to be called very early in bootstrap. That or not cache it, or make it lazy

or leave it as is, and just refer to comments here if asked :)